operator/iamkeycontroller/create.go

```diff
@@ -0,0 +1,146 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"strings"
+)
+
+// Create implements managed.ExternalClient.
+func (p *IAMKeyPipeline) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	log := controllerruntime.LoggerFrom(ctx)
+	log.Info("Creating resource")
+
+	iam := fromManaged(mg)
+	if iam.Status.AtProvider.KeyID != "" {
+		// IAMKey already exists
+		return managed.ExternalCreation{}, nil
+	}
+
+	pipe := pipeline.NewPipeline().WithBeforeHooks(steps.DebugLogger(ctx)).
+		WithSteps(
+			pipeline.NewStepFromFunc("create IAM key", p.createIAMKeyFn(iam)),
+			pipeline.If(hasSecretRef(iam),
+				pipeline.NewStepFromFunc("create credentials secret", p.createCredentialsSecretFn(iam)),
+			),
+			pipeline.NewStepFromFunc("emit event", p.emitCreationEventFn(iam)),
+		)
+	result := pipe.RunWithContext(ctx)
+	if result.IsFailed() {
+		return managed.ExternalCreation{}, errors.Wrap(result.Err(), "cannot create IAM Key")
+	}
+
+	return managed.ExternalCreation{ConnectionDetails: toConnectionDetails(p.exoscaleIAMKey)}, nil
+}
+
+// createIAMKeyFn creates a new IAMKey in the project associated with the API Key and Secret.
+func (p *IAMKeyPipeline) createIAMKeyFn(iamKey *exoscalev1.IAMKey) func(ctx context.Context) error {
+	return func(ctx context.Context) error {
+		exoscaleClient := p.exoscaleClient
+		log := controllerruntime.LoggerFrom(ctx)
+
+		var keyResources []v2.IAMAccessKeyResource
+		for _, bucket := range iamKey.Spec.ForProvider.SOS.Buckets {
+			keyResource := v2.IAMAccessKeyResource{
+				Domain:       SOSResourceDomain,
+				ResourceName: bucket,
+				ResourceType: BucketResourceType,
+			}
+			keyResources = append(keyResources, keyResource)
+		}
+
+		exoscaleIAM, err := exoscaleClient.CreateIAMAccessKey(ctx, iamKey.GetZone(), iamKey.GetKeyName(), v2.CreateIAMAccessKeyWithResources(keyResources))
+		if err != nil {
+			return err
+		}
+
+		if err != nil {
+			return err
+		}
+		// Limitation by crossplane: The interface managed.ExternalClient doesn't allow updating the resource during creation except annotations.
+		// But we need to somehow store the key ID returned by the creation operation, because exoscale API allows multiple IAM Keys with the same display name.
+		// So we store it in an annotation since that is the only allowed place to update our resource.
+		// However, once we observe the spec again, we will copy the key ID from the annotation to the status field,
+		//  and that will become the authoritative source of truth for future reconciliations.
+		metav1.SetMetaDataAnnotation(&iamKey.ObjectMeta, KeyIDAnnotationKey, *exoscaleIAM.Key)
+		metav1.SetMetaDataAnnotation(&iamKey.ObjectMeta, BucketsAnnotationKey, getBuckets(*exoscaleIAM.Resources))
```

Comment:
Why do we save the bucket names in the annotations also? AFAICT once we have the KeyID we can look up which bucket names are allowed during `Observe`.

Reply:
that's a good point, I somehow missed that. I changed the logic and I get the resources from the iamKey in the `observe` step